Persephone/MPDClient/MPDClient.swift

```diff
@@ -12,285 +12,43 @@ import mpdclient
-    case .fetchQueue:
-      sendFetchQueue()
-    case .fetchAllAlbums:
-      allAlbums()
-    }
-  }
-
-  func sendNextTrack() {
-    guard let state = status?.state,
-      state.isOneOf([.playing, .paused])
-      else { return }
-
-    mpd_run_next(connection)
-  }
-
-  func sendPreviousTrack() {
-    guard let state = status?.state,
-      state.isOneOf([.playing, .paused])
-      else { return }
-
-    mpd_run_previous(connection)
-  }
-
-  func sendStop() {
-    mpd_run_stop(connection)
-  }
-
-  func sendPlay() {
-    if status?.state == .stopped {
-      mpd_run_play(connection)
-    } else {
-      mpd_run_toggle_pause(connection)
-    }
-  }
-
-  func sendRunStatus() {
-    guard let status = mpd_run_status(connection) else { return }
-    self.status = Status(status)
-  }
-
-  func sendFetchQueue() {
-    self.queue = []
-    mpd_send_list_queue_meta(connection)
-
-    while let mpdSong = mpd_recv_song(connection) {
-      let song = Song(mpdSong)
-      self.queue.append(song)
-    }
-  }
-
-  func allAlbums() {
-    var albums: [Album] = []
-    var artist: String = ""
-
-    mpd_search_db_tags(connection, MPD_TAG_ALBUM)
-    mpd_search_add_group_tag(connection, MPD_TAG_ALBUM_ARTIST)
-    mpd_search_commit(connection)
-    while let mpdPair = mpd_recv_pair(connection) {
-      let pair = Pair(mpdPair)
-
-      switch pair.name {
-      case "AlbumArtist":
-        artist = pair.value
-      case "Album":
-        albums.append(Album(title: pair.value, artist: artist))
-      default:
-        break
-      }
-
-      mpd_return_pair(connection, pair.mpdPair)
-    }
-
-    delegate?.didLoadAlbums(mpdClient: self, albums: albums)
-  }
-
-  func noIdle() {
-    mpd_send_noidle(connection)
-  }
-
-  func idle() {
-    commandQueue.async { [unowned self] in
-      mpd_send_idle(self.connection)
-
-      let result = mpd_recv_idle(self.connection, true)
-      self.handleIdleResult(result)
-    }
-  }
-
-  func handleIdleResult(_ result: mpd_idle) {
-    let mpdIdle = Idle(rawValue: result.rawValue)
-
-    if mpdIdle.contains(.queue) {
-      self.fetchQueue()
-      self.delegate?.didUpdateQueue(mpdClient: self, queue: self.queue)
-    }
-    if mpdIdle.contains(.player) {
-      self.fetchStatus()
-
-      if let status = self.status {
-        self.delegate?.didUpdateState(mpdClient: self, state: status.state)
-        self.delegate?.didUpdateTime(mpdClient: self, total: status.totalTime, elapsedMs: status.elapsedTimeMs)
-        self.delegate?.didUpdateQueuePos(mpdClient: self, song: status.song)
-      }
-    }
-    if !mpdIdle.isEmpty {
       self.idle()
     }
-  }
-
-  func getLastErrorMessage() -> String? {
-    if mpd_connection_get_error(connection) == MPD_ERROR_SUCCESS {
-      return nil
-    }
-
-    if let errorMessage = mpd_connection_get_error_message(connection) {
-      return String(cString: errorMessage)
-    }
-
-    return nil
+    commandOperation.queuePriority = priority
+    commandsQueued += 1
```

Comment:
I think this might be a data race — it's possible that `queueCommand()` is called while the queue is already executing an operation, meaning that `commandsQueued` could be accessed by multiple threads concurrently.

Reply:
I think you might be right. I wonder what the best solution is? Maybe a stack? Or perhaps a separate serial queue just for these increments/decrements?

Reply:
I’m not sure where you’re using the value of this property, but you might be able to use `OperationQueue`’s own `operationCount` property: https://developer.apple.com/documentation/foundation/operationqueue/1415115-operationcount

Reply:
I started doing this but abandoned it because the value didn't seem to reflect what I was trying to do. Turns out that after I refactored MPDClient to enqueue commands more robustly using this value is now viable.